src/mlpack/core/data/dict_encoding_impl.hpp

```diff
@@ -0,0 +1,234 @@
+/**
+ * @file dict_encoding_impl.hpp
+ * @author Jeffin Sam
+ *
+ * Implementation of dictionary encoding functions.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_DICT_ENCODING_IMPL_HPP
+#define MLPACK_CORE_DATA_DICT_ENCODING_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "dict_encoding.hpp"
+#include <string>
+#include <vector>
+#include <unordered_map>
+
+namespace mlpack {
+namespace data {
+
+  DicitonaryEncoding::DicitonaryEncoding()
+  {
+    // nothing to do here
+  }
+
+  DicitonaryEncoding::~DicitonaryEncoding()
+  {
+    // nothing to do here
+  }
+
+  void DicitonaryEncoding::Reset()
+  {
+    mappings.clear();
+  }
+
+  void DicitonaryEncoding::CreateMap(std::string strings, char deliminator)
+  {
+    std::vector<std::string> temp;
+    boost::split(temp, strings, boost::is_any_of(std::string(1, deliminator)));
+    size_t curLabel = mappings.size() + 1;
+    for (size_t i = 0; i < temp.size(); i++)
+    {
+      mappings[temp[i]] = curLabel;
+      curLabel++;
+    }
+  }
+
+  void DicitonaryEncoding::CreateCharMap(std::string strings)
+  {
+    std::vector<char> temp;
+    std::copy(strings.begin(), strings.end(), std::back_inserter(temp));
+    size_t curLabel = mappings.size();
+    for (size_t i = 0; i < temp.size(); i++)
+    {
+      mappings[(std::string(1, temp[i]))] = curLabel;
+      curLabel++;
+    }
+  }
+
+template<typename eT>
+void DicitonaryEncoding::CharEncode(const std::vector<std::string>& strings,
+            arma::Mat<eT>& output)
+{
+    CharEncode(strings, output, mappings);
+}
+
+template<typename eT>
+void DicitonaryEncoding::CharEncode(const std::vector<std::string>& strings,
+            arma::Mat<eT>& output,
+            std::unordered_map<std::string, size_t>& mapping)
+{
+  std::vector< std::vector<char> > dataset;
+  std::vector<char> temp;
+  size_t globalsize = 0;
+  for (size_t i = 0; i < strings.size(); ++i)
+  {
+    std::copy(strings[i].begin(), strings[i].end(), std::back_inserter(temp));
+    dataset.push_back(temp);
+    globalsize = std::max(globalsize, temp.size());
+    temp.clear();
+  }
+  size_t curLabel = mapping.size() + 1;
+  for (size_t i = 0; i < dataset.size(); ++i)
+  {
+    for (size_t j = 0; j < dataset[i].size(); ++j)
+    {
+      if (mapping.count((std::string(1, dataset[i][j]))) == 0)
+      {
+        mapping[(std::string(1, dataset[i][j]))] = curLabel;
+        ++curLabel;
+      }
+    }
+  }
+  output.zeros(dataset.size(), globalsize);
```

Comment:
May be we could provide a template argument , and leave it to users to choose either mat or sp_mat, because there may be case when the strings are of equal length and thus mat would be fine , but in case where there length are relatively different sp_mat may be good. 

Looking for your take on that?